contrib/pg_tde/src/pg_tde_event_capture.c

```diff
@@ -99,72 +134,50 @@ pg_tde_ddl_command_start_capture(PG_FUNCTION_ARGS)
-		TDEPrincipalKey *principal_key;
+		const char *accessMethod = stmt->accessMethod;
 
 		tdeCurrentCreateEvent.eventType = TDE_TABLE_CREATE_EVENT;
 		tdeCurrentCreateEvent.relation = stmt->relation;
 
-		if (stmt->accessMethod && strcmp(stmt->accessMethod, "tde_heap") == 0)
-		{
-			tdeCurrentCreateEvent.encryptMode = true;
-		}
-		else if ((stmt->accessMethod == NULL || stmt->accessMethod[0] == 0) && strcmp(default_table_access_method, "tde_heap") == 0)
-		{
-			tdeCurrentCreateEvent.encryptMode = true;
-		}
-
-		if (tdeCurrentCreateEvent.encryptMode)
-		{
-			LWLockAcquire(tde_lwlock_enc_keys(), LW_SHARED);
-			principal_key = GetPrincipalKey(MyDatabaseId, LW_SHARED);
-			LWLockRelease(tde_lwlock_enc_keys());
-			if (principal_key == NULL)
-			{
-				ereport(ERROR,
-						(errmsg("failed to retrieve principal key. Create one using pg_tde_set_principal_key before using encrypted tables.")));
+		checkEncryptionClause(accessMethod);
+	}
+	else if (IsA(parsetree, CreateTableAsStmt))
+	{
+		CreateTableAsStmt *stmt = (CreateTableAsStmt *) parsetree;
+		const char *accessMethod = stmt->into->accessMethod;
 
-			}
-		}
+		tdeCurrentCreateEvent.eventType = TDE_TABLE_CREATE_EVENT;
+		tdeCurrentCreateEvent.relation = stmt->into->rel;
 
+		checkEncryptionClause(accessMethod);
 	}
 	else if (IsA(parsetree, AlterTableStmt))
 	{
-		LOCKMODE	lockmode = AccessShareLock; /* TODO. Verify lock mode? */
 		AlterTableStmt *stmt = (AlterTableStmt *) parsetree;
 		ListCell   *lcmd;
 
+		bool		isAccessMethodChange = false;
+
 		foreach(lcmd, stmt->cmds)
 		{
 			AlterTableCmd *cmd = (AlterTableCmd *) lfirst(lcmd);
 
-			if (cmd->subtype == AT_SetAccessMethod &&
-				((cmd->name != NULL && strcmp(cmd->name, "tde_heap") == 0) ||
-				 (cmd->name == NULL && strcmp(default_table_access_method, "tde_heap") == 0))
-				)
+			if (cmd->subtype == AT_SetAccessMethod)
 			{
-				tdeCurrentCreateEvent.encryptMode = true;
+				const char *accessMethod = cmd->name;
+
+				isAccessMethodChange = true;
+
 				tdeCurrentCreateEvent.eventType = TDE_TABLE_CREATE_EVENT;
 				tdeCurrentCreateEvent.relation = stmt->relation;
+				checkEncryptionClause(accessMethod);
 			}
 		}
 
-		if (tdeCurrentCreateEvent.encryptMode)
+		if (EnforceEncryption && isAccessMethodChange && !tdeCurrentCreateEvent.encryptMode)
 		{
-			TDEPrincipalKey *principal_key;
-			Oid			relationId = RangeVarGetRelid(stmt->relation, NoLock, true);
-			Relation	rel = table_open(relationId, lockmode);
-
-			table_close(rel, lockmode);
-
-			LWLockAcquire(tde_lwlock_enc_keys(), LW_SHARED);
-			principal_key = GetPrincipalKey(MyDatabaseId, LW_SHARED);
-			LWLockRelease(tde_lwlock_enc_keys());
-			if (principal_key == NULL)
-			{
-				ereport(ERROR,
-						(errmsg("failed to retrieve principal key. Create one using pg_tde_set_principal_key before using encrypted tables.")));
-
-			}
+			ereport(ERROR,
+					(errmsg("pg_tde.enforce_encryption is ON, only encrypted table access methods are allowed.")));
```

Comment:
This code is unreachable and the error message will always claim that we are creating a table even if we are altering one unless I am mistaken. See the output of the test suite which supports my theory.

Reply:
You are right, I added this section before refactoring the function. Fixed.